mlkem/poly.c

```diff
@@ -299,35 +299,52 @@ void poly_frommsg(poly *r, const uint8_t msg[MLKEM_INDCPA_MSGBYTES]) {
   POLY_BOUND_MSG(r, MLKEM_Q, "poly_frommsg output");
 }
 
-/*************************************************
- * Name:        poly_tomsg
- *
- * Description: Convert polynomial to 32-byte message
- *
- * Arguments:   - uint8_t *msg: pointer to output message
- *              - const poly *a: pointer to input polynomial
- *                  Coefficients must be unsigned canonical
- **************************************************/
 void poly_tomsg(uint8_t msg[MLKEM_INDCPA_MSGBYTES], const poly *a) {
   POLY_UBOUND(a, MLKEM_Q);
 
-  unsigned int i, j;
-  uint32_t t;
-
-  for (i = 0; i < MLKEM_N / 8; i++) {
-    msg[i] = 0;
-    for (j = 0; j < 8; j++) {
-      t = a->coeffs[8 * i + j];
-      // t += ((int16_t)t >> 15) & MLKEM_Q;
-      // t  = (((t << 1) + MLKEM_Q/2)/MLKEM_Q) & 1;
-      t <<= 1;
-      t += 1665;
-      t *= 80635;
-      t >>= 28;
-      t &= 1;
-      msg[i] |= t << j;
+  for (int i = 0; i < MLKEM_N / 8; i++)
+      // clang-format off
+    ASSIGNS(i, OBJECT_WHOLE(msg))
+    INVARIANT(i >= 0 && i <= MLKEM_N / 8)
+    DECREASES(MLKEM_N / 8 - i)
+    // clang-format on
+    {
+      msg[i] = 0;
+      for (int j = 0; j < 8; j++)
+          // clang-format off
+        ASSIGNS(j, OBJECT_WHOLE(msg))
+        INVARIANT(i >= 0 && i <= MLKEM_N / 8 && j >= 0 && j <= 8)
+        DECREASES(8 - j)
```

Comment:
I dropped this clause from newer CBMC proofs because termination is obvious, and I feel it's more useful to minimize the visual burden of the annotations.

Reply:
You're proposing to drop _all_ DECREASES clauses? Or just the ones that are "obvious"? If the latter, please define "obvious"... I agree about reducing the visual clutter. I also agree that:
```
  for (int i = 0; i < C; i++)
  {
    // anything that does not modify i
  }
```
is "obvious"...

Reply:
All obvious ones, which _are_ all, I think? There is no non-trivial termination argument beyond the sampling loops, for which you can only prove that they terminate with probability 1.